Syntax highlight regenerated files

rwp0 opened the pull request:
Such as the regenerated core files ending in: .inc, .tab, and .act

Examples:
https://github.com/Perl/perl5/blob/508aafdf40fc7dd09456ceeddbe19f0fc44e47c9/overload.inc https://github.com/Perl/perl5/blob/508aafdf40fc7dd09456ceeddbe19f0fc44e47c9/perly.act https://github.com/Perl/perl5/blob/508aafdf40fc7dd09456ceeddbe19f0fc44e47c9/perly.tab

![image](https://user-images.githubusercontent.com/68062695/212889688-b148507d-728d-4493-85c0-4e0aab5bc84e.png)

Which are wrongly highlighted.

This will take effect when `regen.pl` is run again.

For `feature.pm` (being a regenerated file) and the like, file version will also need to be increased, I believe.

rwp0 committed:
generated files - update mode lines to specify file type

This updates the mode-line for most of our generated files so that
they include file type information so they will be properly syntax
highlighted on github.

This does not make any other functional changes to the files.

[Note: Commit message rewritten by Yves]

Leont commented:
The vi modeline should probably also be set if you're setting the emacs one?

rwp0 commented:
> The vi modeline should probably also be set if you're setting the emacs one?

That's more for Github highlighting. The Emacs line was already present so I just modified it to include syntax as well. Vi modeline could be added in another PR, I think 🙂

Leont commented:
> Vi modeline could be added in another PR, I think slightly_smiling_face

They are [already](https://github.com/Perl/perl5/blob/42d25076b50038b1cf6cb5cfe41589bc853a0c81/regen/regen_lib.pl#L204) added. Adding a `filetype=c` or `filetype=perl` to that should do the trick.

ilmari commented on regen/regen_lib.pl:
```diff
@@ -130,7 +130,14 @@ sub read_only_top {
 
-    my $raw = "-*- buffer-read-only: t -*-\n";
+    # Generate the "modeline" for syntax highlighting based on the language
+    my $raw;
+    if ($lang eq 'C') {
+        $raw = "-*- mode: C; buffer-read-only: t -*-\n"; # see line 40 where $lang is set: affects .c, .h, .inc, .tab, and .act files
+    }
+    else {
+        $raw = "-*- mode: Perl; buffer-read-only: t -*-\n"; # ibid. ".pod" being the only other one which is in Perl mode as well
+    }
```
Given that `$lang` can only be one of `Perl`, `Pod`, or `C`, and Emacs has `pod-mode`, why not unconditionally add `mode: $lang;`?

rwp0 replied:
Yes, I thought about that as well.

I'm not sure `Pod` will be accepted by GitHub to highlight Pod documents as Linguist calls the language Perl.

I'm not aware if Emacs had pod-mode (natively by default).

Reading [perl-mode](https://github.com/emacs-mirror/emacs/blob/master/lisp/progmodes/perl-mode.el) and [cperl-mode](https://github.com/emacs-mirror/emacs/blob/master/lisp/progmodes/cperl-mode.el) I don't find that easily.

Maybe you mean https://www.emacswiki.org/emacs/PodMode (which might not be in Emacs core, and being available as an extension)?

But if that's really the case I we can do that, yes.

ilmari replied:
GitHub does recognise `-*- mode: Pod -*-` (I just tested by replacing the vim modeline in `INSTALL` on a branch).

You're right, `pod-mode` is not included with Emacs itself, but I expect anyone who's hacking Perl using Emacs and cares about syntax highlighting to have it installed to get highlighting in `.pod` files. I already had it because Debian's `emacs-goodies-el` pulls it in.

rwp0 replied:
Good, I will modify the code to be shorter then.

rwp0 commented:
> > Vi modeline could be added in another PR, I think slightly_smiling_face
> 
> They are [already](https://github.com/Perl/perl5/blob/42d25076b50038b1cf6cb5cfe41589bc853a0c81/regen/regen_lib.pl#L204) added. Adding a `filetype=c` or `filetype=perl` to that (before the final colon) should do the trick.

I didn't notice it in the bottom.

Will add the below code:

```vimrc
ex: set ro ft=c:
ex: set ro ft=perl:
```

rwp0 commented:
> > Such as the regenerated core files ending in: .inc, .tab, and .act
> > Examples: https://github.com/Perl/perl5/blob/508aafdf40fc7dd09456ceeddbe19f0fc44e47c9/overload.inc https://github.com/Perl/perl5/blob/508aafdf40fc7dd09456ceeddbe19f0fc44e47c9/perly.act https://github.com/Perl/perl5/blob/508aafdf40fc7dd09456ceeddbe19f0fc44e47c9/perly.tab
> > ![image](https://user-images.githubusercontent.com/68062695/212889688-b148507d-728d-4493-85c0-4e0aab5bc84e.png)
> > Which are wrongly highlighted.
> > This will take effect when `regen.pl` is run again.
> 
> Speaking of regeneration, t/porting/regen.t is failing in CI tests of your pull request. Could you resolve the problem and re-push?

Yes, I can see that now.

`sh Configure -des -Dusedevel && make test_porting` ([note](https://gist.github.com/rwp0/1ce0f1985abb0b2c1eb1995c2a35c1b6))

Suggestions, please.

```TAP
Test Summary Report
-------------------
porting/authors.t       (Wstat: 768 (exited 3) Tests: 7 Failed: 3)
  Failed tests:  5-7
  Non-zero exit status: 3
porting/regen.t         (Wstat: 65280 (exited 255) Tests: 10 Failed: 1)
  Failed test:  10
  Non-zero exit status: 255
  Parse errors: Bad plan.  You planned 44 tests but ran 10.
porting/utils.t         (Wstat: 0 Tests: 84 Failed: 1)
  Failed test:  39
Files=33, Tests=43727, 63 wallclock secs ( 4.17 usr  0.16 sys + 58.79 cusr  3.78 csys = 66.90 CPU)
Result: FAIL
Finished test run at Tue Jan 17 18:07:07 2023.
make: *** [makefile:877: test_porting] Error 5
```

ilmari commented:
> > Speaking of regeneration, t/porting/regen.t is failing in CI tests of your pull request. Could you resolve the problem and re-push?
> 
> Yes, I can see that now.
> 
> `sh Configure -des -Dusedevel && make test_porting` 
> 
> Suggestions, please.

`make regen regen_perly` should do the trick.

jkeenan commented:
> > 
> > Speaking of regeneration, t/porting/regen.t is failing in CI tests of your pull request. Could you resolve the problem and re-push?
> 
> Yes, I can see that now.
> 
> `sh Configure -des -Dusedevel && make test_porting` ([note](https://gist.github.com/rwp0/1ce0f1985abb0b2c1eb1995c2a35c1b6))
> 
> Suggestions, please.
> 
> ```
> Test Summary Report
> -------------------
> porting/authors.t       (Wstat: 768 (exited 3) Tests: 7 Failed: 3)
>   Failed tests:  5-7
>   Non-zero exit status: 3
> porting/regen.t         (Wstat: 65280 (exited 255) Tests: 10 Failed: 1)
>   Failed test:  10
>   Non-zero exit status: 255
>   Parse errors: Bad plan.  You planned 44 tests but ran 10.
> porting/utils.t         (Wstat: 0 Tests: 84 Failed: 1)
>   Failed test:  39
> Files=33, Tests=43727, 63 wallclock secs ( 4.17 usr  0.16 sys + 58.79 cusr  3.78 csys = 66.90 CPU)
> Result: FAIL
> Finished test run at Tue Jan 17 18:07:07 2023.
> make: *** [makefile:877: test_porting] Error 5
> ```

Running `make test_porting` on your p.r. right now, I'm only seeing failures in `t/porting/regen.t`.  If you run that test file through the harness by itself, the output will provide you with a hint as to how to proceed.

rwp0 commented:
> > > Speaking of regeneration, t/porting/regen.t is failing in CI tests of your pull request. Could you resolve the problem and re-push?
> > 
> > 
> > Yes, I can see that now.
> > `sh Configure -des -Dusedevel && make test_porting`
> > Suggestions, please.
> 
> For the `regen.t` failure, `make regen regen_perly` should do the trick.

`regen_perly` doesn't seem to accept Bison above 3.7 (mine is 3.8.2 under Ubuntu 22.04) :-)

Tux commented:
> > > Vi modeline could be added in another PR, I think slightly_smiling_face
> > 
> > 
> > They are [already](https://github.com/Perl/perl5/blob/42d25076b50038b1cf6cb5cfe41589bc853a0c81/regen/regen_lib.pl#L204) added. Adding a `filetype=c` or `filetype=perl` to that (before the final colon) should do the trick.
> 
> Right, I didn't notice the vi modeline at the bottom of files.
> 
> Will add the below code:
> 
> ```viml
> ex: set ro ft=c:
> ex: set ro ft=perl:
> ```

This will **FAIL** on non-vim vi-clones like elvis: "bad option name ft".

iabyn commented:
On Tue, Jan 17, 2023 at 06:44:45AM -0800, Elvin Aslanov wrote:
> `regen_perly` doesn't seem to accept Bison above 3.7 (mine is 3.8.2 under Ubuntu 22.04) :-)

regen_perly is designed to complain if handed a bison version it doesn't
recognise. This is meant as a precaution in case of incompatible changes,
especially as the script modifies the files generated by bison. Usually if
I find myself with a newer version of bison, I run regen_perly modified to
temporarily accept the newer version, then see if there are any
significant changes in the generated files. If not, I commit the modified
regen_perly with the updated allowed version list.

-- 
"But Sidley Park is already a picture, and a most amiable picture too.
The slopes are green and gentle. The trees are companionably grouped at
intervals that show them to advantage. The rill is a serpentine ribbon
unwound from the lake peaceably contained by meadows on which the right
amount of sheep are tastefully arranged." -- Lady Croom, "Arcadia"

demerphq commented:
FTFY @rwp0

demerphq commented:
Oh man, the regen tests in porting are deeply annoying.

rwp0 commented:
Right, I'm stuck.

If can deal with tests, I can push the other changes as well.

Namely, including vim, and placing the variable directly within string.

demerphq commented:
@rwp0 - im on it. this is a *perfect* PR to make t/porting/regen.t not suck. leave it to me from here.

demerphq committed:
regen/regen_lib.pl - don't leave -new files in the tree when tests fail

Unless the user requests we do so by setting REGEN_T_KEEP_CHANGES

demerphq committed:
t/porting/regen.t - no more "whack-a-mole" subtest execution

Prior to this patch t/porting/regen.t would stop after it found a
single case where the generated file was not up to date. If there
were many files this meant that you would fix, run the test again
find out about the next one, and repeat over and over... This
kind of test whack-a-mole is very frustrating.  To make things worse
the old hint when there was a failure would tell you to run make regen,
which was only true for certain of the tests that might have failed.

This patch updates this logic so we test ALL the files in one go,
and then if there were any issues we die with a report detailing
*all* of the regen commands that need to be run.

It also includes some minor cleanup and tweaks so it runs properly
when executed from the root dir of the repo.

demerphq commented:
@rwp0 hopefully it should pass test now.

rwp0 commented:
> @rwp0 hopefully it should pass test now.

Thanks.

Should I commit the two other changes on top of it?

> Namely, including vim, and placing the variable directly within string.

demerphq commented:
Which other two changes? I guess so. You or I might have to regen again.

demerphq commented:
I'm not sure what you did there in that last patch. It looks like maybe you perltidied regen/regen_lib.pl. Normally we don't do that kind of thing  and follow the style of the file (whatever it was). do you think you can reverse that out and just make the change you needed to make?

rwp0 commented:
I added the changes, seems like the tests fail again.

Even though I did:

    ./Configure -des && make
    ./perl -Ilib t/porting/regen.t=
    # followed the output commands
    # such as  /home/regular/IdeaProjects/perl5/perl -Ilib regen/mk_PL_charclass.pl
    make regen

rwp0 commented:
This should pass this time.

@demerphq please squash the commits for it to be merge-able, or otherwise do what's needed.

Thanks for your support with this.

rwp0 commented:
Apparently I wasn't in `.mailmap`, only in `AUTHORS`, and running the `updateAUTHORS.pl` script added me into that file as well.

Not sure what's the difference between the two, and why my name was missing in the former if that's also required for checks.

rwp0 commented:
NB. this didn't modify `perly.act` and `perly.tab` unlike `overload.inc`

demerphq commented:
> Apparently I wasn't in .mailmap, only in AUTHORS, and running the updateAUTHORS.pl script added me into that file as well.

Actually you added a new email address, and the tests detected it and updateAUTHORS.pl created the .mailmap entry so they both will show your "canonical" details on github and normal git log usage. You can see it by use --no-use-mailmap in git log.

```
$ git log --pretty="short" -2 --no-use-mailmap c7b1281c0d1791f40db04c34e9360dabab377c18
commit c7b1281c0d1791f40db04c34e9360dabab377c18
Author: rwp0 <rwp.primary@gmail.com>

    Add vi modeline

commit 369c52f187042400da20de7f54c10c98374b538c
Author: Elvin Aslanov <rwp.primary@gmail.com>

    Syntax highlight regenerated files
```

> NB. this didn't modify perly.act and perly.tab unlike overload.inc

OK, Ill take a look and probably fix it for you. I assume we don't test that these files are up to date, so we didn't notice that they havent been modified. This is true for a variety of files which are generated but where the generation and testing would be prohibitively slow.

demerphq committed:
perly.* - update generated files to include file type in modeline

This is a companion patch to the previous one, it is generated with a
slightly more modern Bison which produces better indented output, and
puts parens around constant defines and other minor changes. It is
recommended to view this patch with the -w option to skip most of the
the "non change" changes it contains.

demerphq committed:
manually triggered generated files - add file type data to modeline

so that github syntax highlights them properly

rwp0 commented:
Thanks @demerphq for taking this draft PR over and finishing it quickly.

You're really awesome! 🥳

demerphq commented:
No problem.  Happy to help.
